Expand on CREATE_RELATIONSHIP and DELETE_RELATIONSHIP rules (#310)
neo4j/docs-graphql: modules/ROOT/pages/security/authorization.adoc

]) {
    title: String!
    content: String!
    author: User @relationship(type: "AUTHORED", direction: IN)
}
----

=== A note on CREATE_RELATIONSHIP and DELETE_RELATIONSHIP

> Creating or deleting a relationship targets both ends of the relationship, therefore the `CREATE_RELATIONSHIP` and `DELETE_RELATIONSHIP` validation rules will run for both target and source node.

Reviewer: ```suggestion
Creating or deleting a relationship affects both ends of the relationship.
This means that the `CREATE_RELATIONSHIP` and `DELETE_RELATIONSHIP` validation rules apply to both the source and the target node.
```

Author: I'm not sure about this change. It doesn't have to account to both ends, it simply does

Reviewer: how about "affects"?

> Note that for `CREATE_RELATIONSHIP` only the `AFTER` validation rule will run for the source node.

Reviewer: ```suggestion
Note that for `CREATE_RELATIONSHIP` only the `AFTER` validation applies for the source node.
```

> For example, consider a `Post` type and an `User` type both with CREATE_RELATIONSHIP validation rules for BEFORE and AFTER. 

Reviewer: ```suggestion
For example, consider a `Post` type and a `User` type both with CREATE_RELATIONSHIP validation rules for BEFORE and AFTER:
```

* only an existing author of the post can add another co-author. 

> * only a user which is not locked can be added as an author to a post.

Reviewer: ```suggestion
* Only an existing author of the post can add another co-author. 
* Only a user which is not locked can be added as an author to a post.
* 

```

> The rules for each type can be combined into one but we are separating them here for clarity:

Reviewer: ```suggestion
You can combine the rules for each type to a single rule statement.
The following example separates them for clarity:
```
how can it be combined? not saying that we should do both, but a hint i think makes it more readable

Author: It's mentioned just in the sub-section before this one.

Reviewer: `operations: [CREATE_RELATIONSHIP], when: [BEFORE, AFTER]` <- like this?

alright, i think we don't need to add that then :)

Author: yes, exactly :D

> [source, graphql, indent=0]
> ----
> type Post @node @authorization(validate: [
>     { operations: [CREATE_RELATIONSHIP], when: [BEFORE], where: { node: { authors: { some: { id: { eq: "$jwt.sub" } } } } } }  # rule 1
>     { operations: [CREATE_RELATIONSHIP], when: [AFTER], where: { node: { authors: { some: { id: { eq: "$jwt.sub" } } } } } }   # rule 2
> ]) {
>     title: String!
>     content: String!
>     authors: [User!]! @relationship(type: "AUTHORED", direction: IN)
> }
> type User @node @authorization(validate: [
>     { operations: [CREATE_RELATIONSHIP], when: [BEFORE], where: { node: { locked: { eq: false } } } }               # rule 3
>     { operations: [CREATE_RELATIONSHIP], when: [AFTER], where: { node: { locked: { eq: false } } } }                # rule 4
> ]) {
>     id: ID!
>     locked: Boolean!
> }
> ----

Reviewer: ```suggestion
[source, graphql, indent=0]
----
type Post @node @authorization(validate: [
    { operations: [CREATE_RELATIONSHIP], when: [BEFORE], where: { node: { authors: { some: { id: { eq: "$jwt.sub" } } } } } }  <1>
    { operations: [CREATE_RELATIONSHIP], when: [AFTER], where: { node: { authors: { some: { id: { eq: "$jwt.sub" } } } } } }   <2>
]) {
    title: String!
    content: String!
    authors: [User!]! @relationship(type: "AUTHORED", direction: IN)
}
type User @node @authorization(validate: [
    { operations: [CREATE_RELATIONSHIP], when: [BEFORE], where: { node: { locked: { eq: false } } } }               <3>
    { operations: [CREATE_RELATIONSHIP], when: [AFTER], where: { node: { locked: { eq: false } } } }                <4>
]) {
    id: ID!
    locked: Boolean!
}
----
<1> Rule 1
<2> Rule 2
<3> Rule 3
<4> Rule 4
```

tried to apply callouts (https://docs.asciidoctor.org/asciidoc/latest/verbatim/callouts/)

> When executing the following query, the rules that will run are rules 2, 3, and 4. Rule 1 will not run because the relationship has not been created yet, so there is no author of the post to check the id of:

Reviewer: ```suggestion
When executing the following query, rules 2, 3, and 4 are applied:
```

[source, graphql, indent=0]
----
mutation {
    createPosts(
        input: [{ title: "My Post", content: "Amazing content", authors: { connect: { where: { node: { id: { eq: "some-id" } } } } } }]
    ) {
        posts {
            title
        }
    }
}

> ----

Reviewer: ```suggestion
----

Rule 1 isn't applied because the relationship has not been created yet, so there is no author of the post to check the ID of.
```
